target/linux/ath79/files/drivers/gpio/rb4xx-gpio.c

```diff
@@ -0,0 +1,152 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * GPIO driver for the MikroTik RouterBoard 4xx series
+ *
+ * Copyright (C) 2008-2011 Gabor Juhos <juhosg@openwrt.org>
+ * Copyright (C) 2008 Imre Kaloz <kaloz@openwrt.org>
+ * Copyright (C) 2015 Bert Vermeulen <bert@biot.com>
+ * Copyright (C) 2020 Christopher Hill <ch6574@gmail.com>
+ *
+ * This file was based on the driver for Linux 2.6.22 published by
+ * MikroTik for their RouterBoard 4xx series devices.
+ */
+#include <linux/platform_device.h>
+#include <linux/gpio/driver.h>
+#include <linux/module.h>
+#include <linux/of_device.h>
+
+#include <mfd/rb4xx-cpld.h>
+
+struct rb4xx_gpio {
+	struct rb4xx_cpld *cpld;
+	struct device *dev;
+
+	struct gpio_chip chip;
+	struct mutex lock;	/* CPLD sleeps (it's SPI) so we need locking */
```

Comment:
So this is something I think is ultimately a judgement call, and I do value your thoughts on how I see it.

Reading the gpio driver docs, it says (to me) that if the underlying provider can sleep (and the cpld can) then "you need to lock me", implying that the locking is for the benefit of the `gpio_chip` and therefore the mutex functionally belongs in here.

Reply:
[This](https://www.kernel.org/doc/html/v4.17/driver-api/gpio/legacy.html#gpio-access-that-may-sleep) is the section I was referring to.

Reply:
I'm not sure I read what you mean. (By the way, I recommend always checking the [latest documentation](https://www.kernel.org/doc/html/latest/driver-api/gpio/legacy.html#gpio-access-that-may-sleep) when writing new drivers :)

What this section describes is the usual warning about never calling routines that can sleep from IRQ context. Our gpio driver can sleep, it says so so the rest is up to the higher layers.

Locking as you need to do it here comes from the fact that multiple calls to e.g. gpio_set() can happen concurrently while some of these sleep. Then you have a concurrent access to the memory used by the driver ('values') and these are the accesses that need to be locked. At least that's my current understanding :)

Reply:
And to complement, unless I'm mistaken I don't see any place in the current gpio code that concurrently write-accesses memory, which is why to me the locking belongs to the cpld code since this is where you need to protect write accesses to `cpld->values`.

Reply:
So I ended up doing a full 180 on this, and moved the `values` back into the gpio driver (so it can fully manage that shared data) and instead split the api on the cpld to take the lower 8 gpios vs the magic number 9 for the extra led.

I'm now of the opinion this is the "least worst" option. There is no duplication of state, and there's no extra traffic on the SPI bus - at the cost of a little odd looking api.

The lock stays where it was, but I removed the dubious explanation about the sleeping SPI bus - thanks for the info on that one.

Reply:
In short the "cpld" driver now only deals with the SPI messaging and the CPLD commands, and the "gpio" driver deals with storing gpio state. I quite like that separation of concerns.